modules/web/js/type-lattice/operator-lattice.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * Holds operator lattice to be used for ternary, binary and unary operators
+ * @class OperatorLattice
+ */
+class OperatorLattice {
+
+    /**
+     * Constructor for operator lattice
+     * @memberof OperatorLattice
+     */
+    constructor() {
+        this._unaryLattice = [];
+        this._binaryLattice = [];
+    }
+
+    /**
+     * Initialize operator lattice from json
+     * @param {any} operatorLatticeJson operator lattice
+     * @memberof OperatorLattice
+     */
+    initFromJson(operatorLatticeJson) {
+        const unaryLattice = {};
+        const binaryLattice = {};
+
+        const unaryLatticeJson = operatorLatticeJson.filter(edge => edge.type === 'unary');
+        const binaryLatticeJson = operatorLatticeJson.filter(edge => edge.type === 'binary');
+
+        unaryLatticeJson.forEach((edge) => {
+            if (!edge.visited) {
+                const operator = edge.operator;
+                unaryLattice[operator] = {};
+                if (!edge.visited) {
```

Comment:
The check `if (!edge.visited)` is already done in line 47. No need to do it again right?  
Also we can do an early return,
`if (edge.visited) { return; }`

Reply:
This is a two level check. We are iterating over the same list twice and hence the two checks.

Reply:
um. but then the second if check should be inside the second forEach ne? These two ifs are in the same iteration. :thinking: